### PR TITLE
Fix input and output annotations

### PR DESCRIPTION
**Pull request by lptr**
With Gradle 6.0 we are starting to enforce annotations on all task getters. Non ad-hoc tasks should use annotations in place of the `Task.getInputs()` and `getOutputs()` APIs. Due to these changes with Gradle 6.0 the applying the Protobuf plugin results in the following deprecation warnings being displayed:

```text
   > Warning: Type 'GenerateProtoTask': property 'buildType' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'builtins' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'descriptorPath' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'descriptorSetOptions' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'fileResolver' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'flavors' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'generateDescriptorSet' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'isTest' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'isTestVariant' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'outputSourceDirectorySet' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'plugins' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'sourceFiles' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'sourceSet' is not annotated with an input or output annotation.
   > Warning: Type 'GenerateProtoTask': property 'variant' is not annotated with an input or output annotation.
   > Warning: Type 'ProtobufExtract': property 'destDir' is not annotated with an input or output annotation.
   > Warning: Type 'ProtobufExtract': property 'isTest' is not annotated with an input or output annotation.
```

This PR fixes the problems and the warnings are removed.

I tried to figure out the role of each property as best I could based on the code alone, but I'm not sure I got everything right.

### CODE REVIEW

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 1]
```diff
@@ -345,13 +386,12 @@ public class GenerateProtoTask extends DefaultTask {
   public void addSourceFiles(FileCollection files) {
     checkCanConfig()
     sourceFiles.from(files)
-    // Register the files as input so that Gradle will check their changes for incremental build
-    inputs.files(files).withPathSensitivity(PathSensitivity.RELATIVE)
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: They are now captured by these annotations instead:

https://github.com/google/protobuf-gradle-plugin/pull/351/files#diff-88190c144bef7b71456e5191c71bb2b3R252-R253

[the remark this answers is not in the document]
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: To clarify: Gradle queries the annotated getters automatically before running the task to determine up-to-dateness and cache key.

File: build.gradle [hunk 2]
```diff
@@ -14,6 +14,7 @@ apply plugin: 'groovy'
 apply plugin: 'kotlin'
 apply plugin: 'maven'
 apply plugin: 'signing'
+apply plugin: 'java-gradle-plugin'
```
Comment (wolfs) [inline build.gradle]: I would be nice to use the plugins block, though that would be a separate PR.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 3]
```diff
@@ -98,23 +110,29 @@ public class GenerateProtoTask extends DefaultTask {
...
      * If true, imports are included in the descriptor set, such that it is self-containing.
      *
      * Default: false
      */
+    @Input
     boolean includeImports
   }
 
+  @Nested
   final DescriptorSetOptions descriptorSetOptions = new DescriptorSetOptions()
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: I suppose this is only an input when `generateDescriptorSet` is true.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 4]
```diff
@@ -177,7 +195,11 @@ public class GenerateProtoTask extends DefaultTask {
     checkInitializing()
     Preconditions.checkState(this.outputBaseDir == null, 'outputBaseDir is already set')
     this.outputBaseDir = outputBaseDir
-    outputs.dir outputBaseDir
+  }
+
+  @OutputDirectory
+  String getOutputBaseDir() {
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: It seems like the `outputBaseDir` isn't the actual output, only a base location for the actual outputs. Each plugin would output in a sub directory in there. I suppose for now declaring all of the base dir as an output is fine.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 5]
```diff
@@ -214,38 +236,58 @@ public class GenerateProtoTask extends DefaultTask {
...
     Preconditions.checkState(!Utils.isAndroidProject(project),
         'sourceSet should not be used in an Android project')
     Preconditions.checkNotNull(sourceSet, 'sourceSet is not set')
     return sourceSet
   }
 
+  @Nullable
+  @Optional
+  @Input
+  String getSourceSetName() {
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Why is the name of the source set an input?

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 6]
```diff
@@ -214,38 +236,58 @@ public class GenerateProtoTask extends DefaultTask {
...
+    return sourceSet?.name
+  }
+
+  @InputFiles
+  @PathSensitive(PathSensitivity.RELATIVE)
   FileCollection getSourceFiles() {
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Should those be `@SkipWhenEmpty` as well?

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy [hunk 7]
```diff
@@ -43,21 +49,29 @@ class ProtobufExtract extends DefaultTask {
...
   }
 
+  @Input
   public boolean getIsTest() {
     Preconditions.checkNotNull(isTest)
     return isTest
   }
 
+  @InputFiles
+  @PathSensitive(PathSensitivity.NAME_ONLY)
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: I am not sure about the `NAME_ONLY`. Are you should or is `RELATIVE` safer?

It seems like the task is copying/extracing from the input files. Maybe it should have a copy-spec instead of plain input files instead?

It also seems like archives can be part of the inputFiles.

For now, using `NAME_ONLY` does not change the behavior, since that is what the plugin declared before.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy [hunk 8]
```diff
@@ -43,21 +49,29 @@ class ProtobufExtract extends DefaultTask {
...
+  public ConfigurableFileCollection getInputFiles() {
+    return inputFiles
+  }
+
   @TaskAction
   void extract() {
     destDir.mkdir()
     boolean warningLogged = false
-    inputs.files.each { file ->
+    inputFiles.each { file ->
```
Comment (wolfs) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: Does this break users of the plugin which register the input files with `protobufExtract.inputs.file(...)`? There doesn't seem to be another way how to register input files currently.

**Comment by dcabasson** [comment]
Looking forward to having the protobuf plugin leverage all the new features of Gradle 5+.

A good next step will be to make task configuration avoidance work because the plugin in its current form doesn't leverage that at all.

**Comment by jfbibeau** [comment]
Thank you for this PR! Will be nice to get the protobuf plugin more aligned with Gradle best practices 👍

File: build.gradle (hunk 2 above)
Comment (lptr) [reply build.gradle]: Agree. Let's keep this PR's scope small, though.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy (hunk 3 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Good point.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy (hunk 4 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: I think declaring this as an `@OutputDirectory` is a step in the right direction without changing the behavior. We can have followup PRs to make it more correct later.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy (hunk 5 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Right, that doesn't make sense. I just removed it.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy (hunk 6 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Good point.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 7 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: Exactly. Let's not change behavior where we don't have to for now.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 8 above)
Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: It does, though I don't expect people to try to configure `ProtobufExtract` directly.

**Comment by lptr** [comment]
Hi, can someone from the Protobuf team take a look at this please?

**Comment by rpalcolea** [comment]
Also https://github.com/google/protobuf-gradle-plugin/issues/304 is relevant

**Comment by rpalcolea** [comment]
Hi @zhangkun83 , any chance you could look at this PR? also https://github.com/google/protobuf-gradle-plugin/issues/304 would be nice as we have seen this plugin creating several tasks in configuration phase that might not be needed unless the user intent is to run the proto tasks

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 9]
```diff
@@ -86,6 +99,7 @@ public class GenerateProtoTask extends DefaultTask {
    *
    * Default: false
    */
+  @Internal
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: "Attached to a task property to indicate that the property is not to be taken into account for up-to-date checking."
It probably should be taken into account for up-to-date checking.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 10]
```diff
@@ -173,11 +190,21 @@ public class GenerateProtoTask extends DefaultTask {
     return Integer.MAX_VALUE
   }
 
+  @Optional
+  @Nested
+  DescriptorSetOptions getDescriptorSetOptions() {
+    return generateDescriptorSet ? descriptorSetOptions : null
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: If `generateDescriptorSet == false` and this property is accessed, throwing an exception explaining what went wrong would be more helpful.

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 11]
```diff
@@ -504,6 +550,16 @@ public class GenerateProtoTask extends DefaultTask {
     }
   }
 
+  @Nested
+  protected Collection<PluginOptions> getBuiltinsInternal() {
+    return builtins
+  }
+
+  @Nested
+  protected Collection<PluginOptions> getPluginsInternal() {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: I suppose these two getters for exposing the two containers as properties so that they can be validated, since there public equivalents cannot be called after configuration is finished. A comment to explain that would be helpful.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy (hunk 7 above)
Comment (zhangkun83) [reply src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy]: Please add a TODO here to revisit the correctness of PathSensitivity.

**Comment by lptr** [comment]
@zhangkun83 Do you want me to do anything else before this can be merged?

File: src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy [hunk 12]
```diff
@@ -504,6 +538,31 @@ public class GenerateProtoTask extends DefaultTask {
     }
   }
 
+  /**
+   * Used to expose inputs to Gradle, not to be called directly.
+   */
+  @Optional
+  @Nested
+  protected DescriptorSetOptions getDescriptorSetOptions() {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: Why is it not to be called directly? This is the public accessor for DescriptorSetOptions used in the example from README.

It seems that you cannot make it throw when `generateDescriptorSet == false` as I requested earlier, because Gradle will query this property unconditionally. This is a bit unfortunate as Gradle doesn't have an annotation for "this is public, but don't use it for up-to-date checks".

Comment (lptr) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: I'm not sure what you mean by `throw when generateDescriptorSet == false`. I reverted to the origin public property being `@Internal` (that's Gradle's annotation for "this is public, but don't use it for up-to-date checks").

Comment (zhangkun83) [reply src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy]: If we made accessing `descriptorSetOptions` throw when `generateDescriptorSet == false`, the user would get more meaningful error message (see `getDescriptorPath()`), compared to getting `null` for `descriptorSetOptions` and then NPE down the road.

I find `@Internal` on a public property to be confusing, especially when you add the auxiliary properties like `getDescriptorSetOptionsInternal()` that also have "Internal" in the name. Can you rename them to `*ForUpToDateCheck()`?

**Comment by lptr** [comment]
Hi @zhangkun83!

The problem I'd like to address here is that the Protobuf plugin in its currently released form does not declare inputs and outputs for the tasks it creates correctly. I don't want to pressure you into any particular way wrt handling errors etc. Yet we seem to be locked in a cycle of you trying to explain to me how you want things to be which I then implement not to your liking. Considering the long round-trip times I feel this is not leading to the desired results.

Why don't we change this around, so you could add the annotations in a separate PR, and I can help review it? What matters for me is to wrap this up in a timely manner, so the many Gradle users that currently have issues with the plugin can be helped.

I'd like to invite you to the Gradle Community Slack workspace. Given the popularity of your plugin, it would be great if we had a more direct connection. Once we fixed the input/output handling there are other improvements that we'd like to discuss, too. Would you mind sending me your email address to lorant@gradle.com so that I can send an invite?

I'd also like to offer help face-to-face if that would work for you. I'm in the CET timezone, we can chat on Google Hangout or somesuch if you want.

**Comment by zhangkun83** [comment]
@lptr sorry for the slow turnaround. The team have gone through some changes and I have been taking vacations. You idea is good, so please review #364.

**Comment by lptr** [comment]
This has been taken care of via https://github.com/google/protobuf-gradle-plugin/pull/364.
